# **TASK SLIDE 4 - Practicing Python for ML**
`author : Ulivia Embun (10322015)`

## **Challenge 1 - Tabel Ringkasan Hasil Beberapa Model**

Tantangan pertama adalah membuat tabel untuk merangkum hasil prediksi dari keenam algoritma (LR, LDA, KNN, CART, NB, dan SVM) dalam menentukan kelas Iris. Cara terbaik untuk merangkum hasil ini adalah dengan menyajikan *confusion matrix* untuk setiap algoritma, seperti yang ditampilkan dalam slide.

*Confusion matrix* menunjukkan berapa banyak prediksi yang benar dan yang salah untuk setiap kelas. Baris mewakili label asli (True Label), dan kolom mewakili label prediksi (Predicted Label).

In [35]:
# library
import sys
import scipy
import numpy as np
import matplotlib
import pandas as pd
import sklearn
from pandas import read_csv
from matplotlib import pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix
from sklearn.metrics import accuracy_score
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
from sklearn.naive_bayes import GaussianNB
from sklearn.svm import SVC

In [22]:
# load dataset Iris
url = "https://raw.githubusercontent.com/jbrownlee/Datasets/master/iris.csv"
names = ['sepal-length', 'sepal-width', 'petal-length', 'petal-width', 'class']
dataset = read_csv(url, names=names)

In [23]:
dataset.head()

sepal-length  sepal-width  petal-length  petal-width        class
0           5.1          3.5           1.4          0.2  Iris-setosa
1           4.9          3.0           1.4          0.2  Iris-setosa
2           4.7          3.2           1.3          0.2  Iris-setosa
3           4.6          3.1           1.5          0.2  Iris-setosa
4           5.0          3.6           1.4          0.2  Iris-setosa

In [24]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   sepal-length  150 non-null    float64
 1   sepal-width   150 non-null    float64
 2   petal-length  150 non-null    float64
 3   petal-width   150 non-null    float64
 4   class         150 non-null    object 
dtypes: float64(4), object(1)
memory usage: 6.0+ KB


In [25]:
# cek distribusi class
print(dataset.groupby('class').size())

class
Iris-setosa        50
Iris-versicolor    50
Iris-virginica     50
dtype: int64


In [26]:
# splitting data
array = dataset.values
X = array[:,0:4]  # kolom 0 sampai 3 untuk fitur
y = array[:,4]    # kolom 4 untuk target

# splitting data 80% training dan 20% testing
X_train, X_validation, Y_train, Y_validation = train_test_split(X, y, test_size=0.20, random_state=1, shuffle=True) #

In [13]:
# define model dari 6 algoritma yang akan dievaluasi
models = []
models.append(('LR', LogisticRegression(solver='liblinear', multi_class='ovr'))) #
models.append(('LDA', LinearDiscriminantAnalysis())) #
models.append(('KNN', KNeighborsClassifier())) #
models.append(('CART', DecisionTreeClassifier())) #
models.append(('NB', GaussianNB())) #
models.append(('SVM', SVC(gamma='auto'))) #

In [28]:
# prediksi dan evaluasi untuk setiap model
print("\n--- Hasil Evaluasi Model pada Data Validasi ---\n")

for name, model in models:
    # training model with data training
    model.fit(X_train, Y_train)

    # cross check with data testing
    predictions = model.predict(X_validation)

    # print hasil evaluasi
    print(f"========================== {name} ==========================")
    print(f"Accuracy: {accuracy_score(Y_validation, predictions)}")
    print("\nConfusion Matrix:")
    print(confusion_matrix(Y_validation, predictions))


--- Hasil Evaluasi Model pada Data Validasi ---

========================== LR ==========================
Accuracy: 0.8333333333333334

Confusion Matrix:
[[11  0  0]
 [ 0  8  5]
 [ 0  0  6]]
========================== LDA ==========================
Accuracy: 1.0

Confusion Matrix:
[[11  0  0]
 [ 0 13  0]
 [ 0  0  6]]
========================== KNN ==========================
Accuracy: 1.0

Confusion Matrix:
[[11  0  0]
 [ 0 13  0]
 [ 0  0  6]]
========================== CART ==========================
Accuracy: 0.9666666666666667

Confusion Matrix:
[[11  0  0]
 [ 0 12  1]
 [ 0  0  6]]
========================== NB ==========================
Accuracy: 0.9666666666666667

Confusion Matrix:
[[11  0  0]
 [ 0 12  1]
 [ 0  0  6]]
========================== SVM ==========================
Accuracy: 0.9666666666666667

Confusion Matrix:
[[11  0  0]
 [ 0 12  1]
 [ 0  0  6]]


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:1256: FutureWarning: 'multi_class' was deprecated in version 1.5 and will be removed in 1.7. Use OneVsRestClassifier(LogisticRegression(..)) instead. Leave it to its default value to avoid this warning.
  warnings.warn(


**Tabel Rangkuman Hasil (disajikan dengan Confusion Matrix)**

| Algoritma | Confusion Matrix (Kelas: 1. Iris-setosa, 2. Iris-versicolor, 3. Iris-virginica) |
| :--- | :--- |
| **LR** (Logistic Regression) | `[[11, 0, 0], [0, 8, 5], [0, 0, 6]]` |
| **LDA** (Linear Discriminant Analysis) | `[[11, 0, 0], [0, 13, 0], [0, 0, 6]]` |
| **KNN** (K-Nearest Neighbors) | `[[11, 0, 0], [0, 13, 0], [0, 0, 6]]` |
| **CART** (Classification and Regression Trees) | `[[11, 0, 0], [0, 12, 1], [0, 0, 6]]` |
| **NB** (Gaussian Naive Bayes) | `[[11, 0, 0], [0, 12, 1], [0, 0, 6]]` |
| **SVM** (Support Vector Machines) | `[[11, 0, 0], [0, 12, 1], [0, 0, 6]]` |

---

## **Challenge 2 - Perhitungan dan Perbandingan Metrik Evaluasi**

Tantangan kedua dan ketiga meminta untuk menghitung metrik `accuracy`, `precision`, `recall`, dan `F1-score` untuk setiap algoritma dan membandingkannya dengan hasil yang diberikan oleh `sklearn`.

**Perhitungan Manual**

Seperti yang disarankan oleh petunjuk pada slide, untuk menghitung metrik per kelas, kita menggunakan pendekatan "satu lawan semua" (one-vs-rest). Untuk sebuah kelas A, maka:
* **True Positive (TP):** Kelas A yang diprediksi sebagai kelas A.
* **False Negative (FN):** Kelas A yang diprediksi sebagai kelas lain.
* **False Positive (FP):** Kelas lain yang diprediksi sebagai kelas A.
* **True Negative (TN):** Kelas lain yang diprediksi sebagai kelas lain.

Formula yang digunakan adalah sebagai berikut:
* $Accuracy = \frac{TP+TN}{TP+TN+FP+FN}$
* $Precision = \frac{TP}{TP+FP}$
* $Recall = \frac{TP}{TP+FN}$
* $F1-score = 2 \cdot \frac{precision \cdot recall}{precision+recall}$

In [38]:
# perhitungan hasil manual dengan formula di slide
def hitung_metrik_manual(model_name, cm):
    """
    Menghitung metrik evaluasi dari confusion matrix dan mengembalikannya
    sebagai dictionary.
    """
    class_labels = ['Iris-setosa', 'Iris-versicolor', 'Iris-virginica']
    total_samples = np.sum(cm)
    results = {'Model': model_name}

    # Hitung akurasi keseluruhan
    accuracy = np.sum(np.diag(cm)) / total_samples
    results['Overall Accuracy'] = f"{accuracy:.2f}"

    # Hitung metrik untuk setiap kelas
    for i, label in enumerate(class_labels):
        tp = cm[i, i]
        fp = np.sum(cm[:, i]) - tp
        fn = np.sum(cm[i, :]) - tp

        precision = tp / (tp + fp + 1e-8)
        recall = tp / (tp + fn + 1e-8)
        f1_score = 2 * (precision * recall) / (precision + recall + 1e-8)

        results[f'Precision ({label})'] = f"{precision:.2f}"
        results[f'Recall ({label})'] = f"{recall:.2f}"
        results[f'F1-Score ({label})'] = f"{f1_score:.2f}"

    return results

# Kumpulan Confusion Matrix dari Semua Model
all_models_cm = {
    'LR': np.array([[11, 0, 0], [0, 8, 5], [0, 0, 6]]),
    'LDA': np.array([[11, 0, 0], [0, 13, 0], [0, 0, 6]]),
    'KNN': np.array([[11, 0, 0], [0, 13, 0], [0, 0, 6]]),
    'CART': np.array([[11, 0, 0], [0, 12, 1], [0, 0, 6]]),
    'NB': np.array([[11, 0, 0], [0, 12, 1], [0, 0, 6]]),
    'SVM': np.array([[11, 0, 0], [0, 12, 1], [0, 0, 6]])
}

# Proses semua model dan kumpulkan hasilnya
summary_results = []
for name, cm in all_models_cm.items():
    model_metrics = hitung_metrik_manual(name, cm)
    summary_results.append(model_metrics)

# Buat DataFrame pandas dari hasil
df_summary = pd.DataFrame(summary_results)

# Atur urutan kolom agar lebih mudah dibaca
df_summary = df_summary.set_index('Model')

# Tampilkan DataFrame sebagai tabel yang diformat dengan baik
df_summary

Overall Accuracy Precision (Iris-setosa) Recall (Iris-setosa)  \
Model                                                                 
LR                0.83                    1.00                 1.00   
LDA               1.00                    1.00                 1.00   
KNN               1.00                    1.00                 1.00   
CART              0.97                    1.00                 1.00   
NB                0.97                    1.00                 1.00   
SVM               0.97                    1.00                 1.00   

      F1-Score (Iris-setosa) Precision (Iris-versicolor)  \
Model                                                      
LR                      1.00                        1.00   
LDA                     1.00                        1.00   
KNN                     1.00                        1.00   
CART                    1.00                        1.00   
NB                      1.00                        1.00   
SVM                     1.00                        1.00   

      Recall (Iris-versicolor) F1-Score (Iris-versicolor)  \
Model                                                       
LR                        0.62                       0.76   
LDA                       1.00                       1.00   
KNN                       1.00                       1.00   
CART                      0.92                       0.96   
NB                        0.92                       0.96   
SVM                       0.92                       0.96   

      Precision (Iris-virginica) Recall (Iris-virginica)  \
Model                                                      
LR                          0.55                    1.00   
LDA                         1.00                    1.00   
KNN                         1.00                    1.00   
CART                        0.86                    1.00   
NB                          0.86                    1.00   
SVM                         0.86                    1.00   

      F1-Score (Iris-virginica)  
Model                            
LR                         0.71  
LDA                        1.00  
KNN                        1.00  
CART                       0.92  
NB                         0.92  
SVM                        0.92

**Perhitungan dengan `sklearn`**

Hasil perhitungan manual menggunakan metodologi di atas akan sama persis dengan laporan klasifikasi (`classification_report`) yang dihasilkan oleh pustaka `sklearn`.

In [30]:
# perhitungan hasil dengan sklearn
print("\n--- Hasil Evaluasi Model pada Data Validasi ---\n")

for name, model in models:
    # training model with data training
    model.fit(X_train, Y_train)

    # cross check with data testing
    predictions = model.predict(X_validation)

    # print hasil evaluasi
    print(f"========================== {name} ==========================")
    print("\nClassification Report:")
    print(classification_report(Y_validation, predictions)) # ini sklearn


--- Hasil Evaluasi Model pada Data Validasi ---

========================== LR ==========================

Classification Report:
                 precision    recall  f1-score   support

    Iris-setosa       1.00      1.00      1.00        11
Iris-versicolor       1.00      0.62      0.76        13
 Iris-virginica       0.55      1.00      0.71         6

       accuracy                           0.83        30
      macro avg       0.85      0.87      0.82        30
   weighted avg       0.91      0.83      0.84        30

========================== LDA ==========================

Classification Report:
                 precision    recall  f1-score   support

    Iris-setosa       1.00      1.00      1.00        11
Iris-versicolor       1.00      1.00      1.00        13
 Iris-virginica       1.00      1.00      1.00         6

       accuracy                           1.00        30
      macro avg       1.00      1.00      1.00        30
   weighted avg       1.00      1.00     

/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:1256: FutureWarning: 'multi_class' was deprecated in version 1.5 and will be removed in 1.7. Use OneVsRestClassifier(LogisticRegression(..)) instead. Leave it to its default value to avoid this warning.
  warnings.warn(




Berikut adalah tabel komprehensif yang merangkum metrik evaluasi untuk setiap algoritma, seperti yang dilaporkan oleh `sklearn` pada cell di atas.


In [37]:
sklearn_results = [
    {
        'Model': 'LR', 'Overall Accuracy': 0.83,
        'Precision (Iris-setosa)': 1.00, 'Recall (Iris-setosa)': 1.00, 'F1-Score (Iris-setosa)': 1.00,
        'Precision (Iris-versicolor)': 1.00, 'Recall (Iris-versicolor)': 0.62, 'F1-Score (Iris-versicolor)': 0.76,
        'Precision (Iris-virginica)': 0.55, 'Recall (Iris-virginica)': 1.00, 'F1-Score (Iris-virginica)': 0.71
    },
    {
        'Model': 'LDA', 'Overall Accuracy': 1.00,
        'Precision (Iris-setosa)': 1.00, 'Recall (Iris-setosa)': 1.00, 'F1-Score (Iris-setosa)': 1.00,
        'Precision (Iris-versicolor)': 1.00, 'Recall (Iris-versicolor)': 1.00, 'F1-Score (Iris-versicolor)': 1.00,
        'Precision (Iris-virginica)': 1.00, 'Recall (Iris-virginica)': 1.00, 'F1-Score (Iris-virginica)': 1.00
    },
    {
        'Model': 'KNN', 'Overall Accuracy': 1.00,
        'Precision (Iris-setosa)': 1.00, 'Recall (Iris-setosa)': 1.00, 'F1-Score (Iris-setosa)': 1.00,
        'Precision (Iris-versicolor)': 1.00, 'Recall (Iris-versicolor)': 1.00, 'F1-Score (Iris-versicolor)': 1.00,
        'Precision (Iris-virginica)': 1.00, 'Recall (Iris-virginica)': 1.00, 'F1-Score (Iris-virginica)': 1.00
    },
    {
        'Model': 'CART', 'Overall Accuracy': 0.97,
        'Precision (Iris-setosa)': 1.00, 'Recall (Iris-setosa)': 1.00, 'F1-Score (Iris-setosa)': 1.00,
        'Precision (Iris-versicolor)': 1.00, 'Recall (Iris-versicolor)': 0.92, 'F1-Score (Iris-versicolor)': 0.96,
        'Precision (Iris-virginica)': 0.86, 'Recall (Iris-virginica)': 1.00, 'F1-Score (Iris-virginica)': 0.92
    },
    {
        'Model': 'NB', 'Overall Accuracy': 0.97,
        'Precision (Iris-setosa)': 1.00, 'Recall (Iris-setosa)': 1.00, 'F1-Score (Iris-setosa)': 1.00,
        'Precision (Iris-versicolor)': 1.00, 'Recall (Iris-versicolor)': 0.92, 'F1-Score (Iris-versicolor)': 0.96,
        'Precision (Iris-virginica)': 0.86, 'Recall (Iris-virginica)': 1.00, 'F1-Score (Iris-virginica)': 0.92
    },
    {
        'Model': 'SVM', 'Overall Accuracy': 0.97,
        'Precision (Iris-setosa)': 1.00, 'Recall (Iris-setosa)': 1.00, 'F1-Score (Iris-setosa)': 1.00,
        'Precision (Iris-versicolor)': 1.00, 'Recall (Iris-versicolor)': 0.92, 'F1-Score (Iris-versicolor)': 0.96,
        'Precision (Iris-virginica)': 0.86, 'Recall (Iris-virginica)': 1.00, 'F1-Score (Iris-virginica)': 0.92
    }
]

# Buat DataFrame dan atur 'Model' sebagai index
df_sklearn = pd.DataFrame(sklearn_results).set_index('Model')

# Tampilkan DataFrame
df_sklearn

Overall Accuracy  Precision (Iris-setosa)  Recall (Iris-setosa)  \
Model                                                                    
LR                 0.83                      1.0                   1.0   
LDA                1.00                      1.0                   1.0   
KNN                1.00                      1.0                   1.0   
CART               0.97                      1.0                   1.0   
NB                 0.97                      1.0                   1.0   
SVM                0.97                      1.0                   1.0   

       F1-Score (Iris-setosa)  Precision (Iris-versicolor)  \
Model                                                        
LR                        1.0                          1.0   
LDA                       1.0                          1.0   
KNN                       1.0                          1.0   
CART                      1.0                          1.0   
NB                        1.0                          1.0   
SVM                       1.0                          1.0   

       Recall (Iris-versicolor)  F1-Score (Iris-versicolor)  \
Model                                                         
LR                         0.62                        0.76   
LDA                        1.00                        1.00   
KNN                        1.00                        1.00   
CART                       0.92                        0.96   
NB                         0.92                        0.96   
SVM                        0.92                        0.96   

       Precision (Iris-virginica)  Recall (Iris-virginica)  \
Model                                                        
LR                           0.55                      1.0   
LDA                          1.00                      1.0   
KNN                          1.00                      1.0   
CART                         0.86                      1.0   
NB                           0.86                      1.0   
SVM                          0.86                      1.0   

       F1-Score (Iris-virginica)  
Model                             
LR                          0.71  
LDA                         1.00  
KNN                         1.00  
CART                        0.92  
NB                          0.92  
SVM                         0.92

## **Kesimpulan Analisis:**

* **Performa Terbaik:** Algoritma **LDA** dan **KNN** mencapai akurasi sempurna (1.00) pada data validasi, memprediksi semua 30 sampel dengan benar.
* **Performa Sangat Baik:** Algoritma **CART**, **NB**, dan **SVM** juga menunjukkan performa yang sangat kuat dengan akurasi 96.7%. Ketiga model ini hanya salah mengklasifikasikan satu sampel (satu *Iris-versicolor* diprediksi sebagai *Iris-virginica*).
* **Performa Terendah:** Algoritma **Logistic Regression (LR)** memiliki akurasi terendah yaitu 83.3%. Kesalahan utamanya adalah salah mengklasifikasikan lima sampel *Iris-versicolor* sebagai *Iris-virginica*.
* **Perbandingan dengan `sklearn`:** Hasil perhitungan manual menggunakan formulas yang diberikan terbukti **sesuai** dengan nilai-nilai yang disajikan dalam tabel di atas, yang diambil langsung dari laporan `sklearn`. Ini menunjukkan bahwa pemahaman tentang TP, TN, FP, dan FN adalah fundamental untuk menginterpretasikan output dari pustaka machine learning.